### ISSUE TITLE
Server-side rendering

### ISSUE THREAD

**Issue by azaslonov** [knockout/knockout#2214]
Are you guys planning to support server-side rendering in the future? If not, what do you think it might take to implement it and what to look out for?

Thank you in advance.

**Comment by mbest** [comment]
Check out https://www.npmjs.com/package/knockout-pre-rendered

**Comment by azaslonov** [comment]
Thank you sir, but the question was about generating the HTML on the server like do Angular Universal (https://github.com/angular/universal) or Vue (https://vuejs.org/v2/guide/ssr.html).

**Comment by mbest** [comment]
No, Knockout doesn't care what you use on the server to pre-render the HTML.

**Comment by mbest** [comment]
We use [PhantomJS](http://phantomjs.org/) to run Knockout unit-tests and I imagine it could be used to do some server-side rendering.

**Comment by azaslonov** [comment]
Yeah, basically, what I'm trying to achieve is to take same view model classes and KO templates (that I run on the client, with all the bindings) and run them on Node JS server to generate static HTML ready to be sent to the client, where it can be used with "knockout-pre-rendered" to hydrate the app. They call it "isomorphic" code.

**Comment by mbest** [comment]
Here's another project that could help: https://github.com/prerender/prerender

**Comment by azaslonov** [comment]
Thank you Michael! That may work out as initial approach.

**Comment by kaheglar** [comment]
You could try something like this https://gist.github.com/kaheglar/10afd535e4970ad9ec6accd62b6ceb03.

**Comment by azaslonov** [comment]
Stephen, I think this is exactly what I need! Thank you so much!

**Comment by iberodev** [comment]
@azaslonov did you have a good result with knockout and domino for server side rendering?
We're also looking for some option like Angular Universal but for knockoutJs, Thanks

**Comment by azaslonov** [comment]
hi @iberodev, yes, we were able to establish server-side rendering with KO and domino (we use it for page publishing), though I would not call it a good result because obviously KO wasn't designed for this scenario. For example, it's hard to tell when KO finished rendering page/component (at least we couldn't find reliable approach), so you don't know when to return control/response to a caller.

**Comment by kaheglar** [comment]
Hi @azaslonov @iberodev. The trick is to make the asynchronous aspects of your application synchronous, when rendering on the server side. The two main cases are:

1. Async. data loading - Ensure all the necessary state is passed into `applyBindings`.
2. Async. component loading – Preload (require) all the async components before calling `applyBindings`.

`applyBinding` should then return a fully rendered page.

**Comment by azaslonov** [comment]
Hi @kaheglar,
Yes, that was my first thought as well. However, even having the whole model in memory KO creates DOM elements one by one (it's not like React building everything in virtual DOM and then inserts complete fragment into real one), so we still can't say for sure when it's done updating the document.

**Comment by miellaby** [comment]
> KO creates DOM elements one by one
@azaslonov

ko.applyBinding() call is asynchronous? even if component.register with `synchronous: true` ?  
Also, did you try calling ko.tasks.runEarly() ?

**Comment by brianmhunt** [comment]
For [tko](github.com/knockout/tko), `ko.applyBindings` returns a promise that resolves when async elements are done rendering.

For bindings (like `component`) that call `applyBindings`, the binding promise only resolves when the sub-bindings are similarly resolved.

So one can do `ko.applyBindings(...).then(fn)` and have a reasonable expectation that the first round of rendering will have finished.

**Comment by kaheglar** [comment]
@azaslonov @miellaby You can use this pattern to render async. components synchronously https://gist.github.com/kaheglar/a587286db24af724f4e2b31ee4e57eff

**Comment by codymullins** [comment]
@brianmhunt I think that promise should wait until all nested components have finished rendering. It would make more sense to me at least. Or if the method could accept configuration, i.e. 
```
const model = {};
const config = {
   callbackOptions: { deep: true; }
}
ko.applyBindings(model, config).then(response => {..});
```

Also does the reject get fired if there's an error applying bindings?

**Comment by brianmhunt** [comment]
> for that promise to return when all nested components have finished rendering

That's what it does.

> does the reject get fired if there's an error applying bindings

Not yet, but that's the plan.

**Comment by azaslonov** [comment]
@kaheglar, yes, we're doing about the same. Thanks for suggestion.

@brianmhunt, don't know if guys consider it, but in our case it would awesome if we could have some special methods (called by KO runtime) on view model like:
`onInit(): Promise<void>`
`onChange(): Promise<void>`
this way we could ensure each component has loaded all the data and it's done rendering.

**Comment by brianmhunt** [comment]
@azaslonov It's an interesting suggestion.

As a matter of interest, in tko binding handlers no longer have the `init` and `change`.  They are (or become), instead, ES6 classes derived from [BindingHandler](https://github.com/knockout/tko/blob/master/packages/tko.bind/src/BindingHandler.js)

In TKO, you can continue to use the `{init: / update: }` as it will be converted to a [LegacyBindingHandler](https://github.com/knockout/tko/blob/master/packages/tko.bind/src/LegacyBindingHandler.js).  

You can also add a plain function (which essentially becomes the `init`) e.g. 

```javascript
ko.bindingHandlers.set('debugger', function (element, valueAccessor, allBindings, vm, $context) { debugger
})
```

I'm not sure how `onChange` in particular might work, but I think it's certainly possible to negotiate a callback via the context.

My one concern is that this is very “inner-loop”, so I feel there's a risk of some substantial performance issues here.  That said, I think it's worth considering. Please feel free to open an issue on knockout/tko so we can track it.

Cheers

**Comment by nerfpops** [comment]
Since knockout removes whatever html that's nested inside `data-bind="if: ...."`, how would one pick up the server rendered html and have the same client side knockout view model take over? Just curious, I'm very interested in having react-like server rendering in knockout.

**Comment by kaheglar** [comment]
Knockout destroys the server rendered HTML then recreated it (hopefully verbatim) with the applyBindings call. This sounds horrific, but in practice it works quite well. If all the bindings are synchronous, this seamless. Asynchronous components can cause a nasty flicker, but this can be easily rectified if they are pre-loaded (made synchronous).

The other option is this http://www.erikschierboom.com/2015/04/25/knockout-pre-rendered/.

**Comment by chrisknoll** [comment]
So, @nerfpops brings up an intersting case that I thought would be a problem re: the if: binding.  I was just looking at code in my own project that uses the if: binding, and if the if: binding results in FALSE, then the inner content isn't presented in the dom.  This means that you can't take the generated HTML from the server and use it as the template to bind to using applyBindings().  (Ie: the content in the if: binding in the template isn't rendered and therefore the client doesn't know what should have gone into the if: block if the if: bound data changes to something truthy.  

Maybe I'm misunderstanding something, but seems to me that the rendered HTML isn't the same as the HTML template that you bind to, so it can't be a matter of doing the first render on the server, and then applyBinding() to the html on the client.  

One question that strikes me is that when you do have a local template and the if: binding is applied, where does the inner content of the if: binding get saved to when the inner dom elements are removed (due to the if: being false)?  Must be parsed-and-stored somewhere in the knockout pipeline, but the problem is when you start off with a server-rendered template, this missing information seems to make the idea of server-side rendering not possible, at least in the case of the if: binding.

@kaheglar , was there some aspect of the knockout rendering pipeline that i'm missing that would let you applyBindigs() to a server-rendered block of HTML?  I'm not understanding where the inner elemetns of a if: binding would be delivered to the client so that when the if becomes true, ti will generate the necessary dom elements.

Just so I understand the context of the original thread: the react folks assert that you can shave time off of a page's time-to-interact if you render the first data on the server and send down the HTML and then attach the client side rendering in Javascript for user interactivity (so starts rendering on server, and continues updating on the client).  This requires that the logic to render the DOM is contained in a separate javascript download (ala .jsx files).  I think this sorta construct is replicated with knockout components where you have the model and html view provided as the template parameter...so I am wondering if all of this is OK if you wrap your content in components such that downloading the components will give you the 'inner' if: binding logic that is required to continue rendering the page on the client space.

**Comment by nerfpops** [comment]
Ah, of course, if you use knockout components, it's probably not an issue. Now we just need jsx! Would be a killer feature though, if one could first render the page with Razor, then pass it on to node and prerender the html with knockout, then let the client pick up from there.

**Comment by chrisknoll** [comment]
I'll go down on record saying I do not need jsx :).  In my component usage, I commonly will share a model behavior, but have different layouts (re: read-only vs. updatable or tabbed layout vs accordian view, etc).  What i've seen of JSX is that the advantage is that the behavior and layout is tightly coupled, but that's not the coupling I want to have.  Plus, JSX requires transpiling.   So, don't want to derail this conversation about serverside rendering as a Pro-Con argument for JSX, but at the same time I don't want to give the impression that JSX solves this issue.

**Comment by nerfpops** [comment]
You decide the coupling you want with jsx and everybody transpiles these days. There's also great IDE support and TypeScript support for jsx. If you're going to write html in your js, you might as well use jsx. Not saying it would solve this particular issue, but it would enable us to write components in a type safe way that's easier to refactor than strings with html. It was just a wish coupled with server rendering, but it's something that would be great to have.

**Comment by kaheglar** [comment]
@chrisknoll You've, quite rightly, exposed a potential mantrap here, where a template that is inferred from the index.html can be corrupted/altered by server rendering stage. I find the best way to avoid a mantrap is not to step in it. Ensure your templates are explicitly defined—e.g. using the template binding, or better still a component—and you should sidestep this problem.

**Comment by Andrew3c** [comment]
I can still see the issue is opened.
Can I use Server-side rendering in Knockout (no prerendering?). 

If so, what should I do? I can't understand some previous messages. 
Is there any tutorial, or library? Like Angular Universal for Angular?

**Comment by kaheglar** [comment]
No tutorial that I know of. Here's the naive solution https://runkit.com/kaheglar/58d4df9993b2030014579f47, however, things get much more complicated when you introduce async component and data loading...

**Comment by azaslonov** [comment]
KO extensively uses "window" and "document" objects. Therefore, in our solution based on "domino" we have to use "global.window" and "global.document" respectively (about the same as in @kaheglar's example).  However, there is only one global object, so we can work with only one document at a time.

I wonder if it is possible to make KO instance-based, so that one could specify those objects for each instance?

**Comment by mbest** [comment]
Knockout begins like this:

```javascript
(function(undefined){
    // (0, eval)('this') is a robust way of getting a reference to the global object
    // For details, see http://stackoverflow.com/questions/14119988/return-this-0-evalthis/14120023#14120023
    var window = this || (0, eval)('this'),
        document = window['document'],
        navigator = window['navigator'],
```

As you can see, the `window` and `document` objects are retrieved once, so you could initialize multiple Knockout instances against multiple `document` objects.

Additionally Knockout supports binding against a different `document` object than it was initialized with, although certain things will also use the initialization `document`.

**Comment by tscpp** [comment]
I have made a proof of concept of a simple server-side renderer for knockout. I have the details of how it works written in the project's readme. It needs no simulation to run, but instead relies on isomorphic viewmodels, or viewmodels written specifically for the server-side, with the same underlying data model (it is not as complicated as it sounds, see readme).

As @chrisknoll pointed out, modifying the view server-side for some bindings (if, foreach, etc.) can be destructive. I do not have a solution for this yet, however I plan on writing dynamically generated hydration code specific to that view. This also has the potential to speed up the application a bit as it removes the need for parsing or traversing the document runtime.

This is not a silver bullet that can magically server-side render your whole application without a bit of tweaking. However, I find the solution the best as you can gradually add SSR to the application, especially SEO critical parts of the applications, e.g. articles or documentation.

If anyone is still interested:

https://github.com/tscpp/knockout-ssr
